schema/impl.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schema
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/heapster/sinks/cache"
+)
+
+var lock sync.RWMutex
+
+func NewCluster() Cluster {
+	return newRealCluster()
+}
+
+func newRealCluster() *realCluster {
+	cinfo := ClusterInfo{
+		InfoType:   newInfoType(nil, nil),
+		Namespaces: make(map[string]*NamespaceInfo),
+		Nodes:      make(map[string]*NodeInfo),
+	}
+	cluster := &realCluster{
+		timestamp:   time.Time{},
+		ClusterInfo: cinfo,
+	}
+	return cluster
+}
+
+func (rc *realCluster) GetAllClusterData() (*ClusterInfo, time.Time, error) {
```

Comment:
My initial thinking was to have two methods for each exposable resource, "All" and "New". I don't believe there is any reason to explore field-specific interfaces, at least not at this stage.
GetAllXData() -> returns all metrics and structure under X
GetNewXData(start time.Time) -> returns structure under X and metrics with timestamp > start

"Get" functions return the cluster timestamp at the time of the request. Therefore, the user will be able to get only fresh data by calling the GetNewX methods with that timestamp as a parameter. This can prove useful for AJAX calls

Note: The timestamp signifies the latest metric "end" time, as provided by cadvisor, rather than time.Now(). This allows us to skip timezone conversions and calculations of the time difference from metric generation on cadvisor, until consumption through the heapster API